src/Symfony/Component/Security/Core/User/UserProviderInterface.php

```diff
@@ -28,6 +28,8 @@
  * information is submitted or what the UserInterface object looks like.
  *
  * @author Fabien Potencier <fabien@symfony.com>
+ *
+ * @template TUser of UserInterface
```

Comment:
this should be `@template-covariant` to be more flexible as the template type is only used in covariant positions (once the usage in `supportsClass` is removed)

Reply:
What about adding
```
@param TUser $user
```
to the refreshUser method ?

Because we're supposed to call
```
$provider->supportsClass($userClass)
```
before `refreshUser`.

Reply:
This would forbid making the generic type covariant, which would make it harder to properly write types for code using UserProviderInterface instances.

Reply:
having `UserProviderInterface` being covariant is a big win (it would be the only way to actually fully type ChainUserProvider with static analysis tools at their strict levels).

The template type of EntityUserProvider can be a non-generic one.